quill-core/src/main/scala/io/getquill/quotation/Parsing.scala

```diff
@@ -45,26 +47,48 @@ trait Parsing extends SchemaConfigParsing {
 
     case q"$i: $typ"                        => astParser(i)
 
-    case q"$tupleTree match { case ($fieldsTrees) => $bodyTree }" =>
-      val tuple = astParser(tupleTree)
-      val fields = astParser(fieldsTrees)
-      val body = astParser(bodyTree)
-      def property(path: List[Int]) =
-        path.foldLeft(tuple) {
-          case (t, i) => Property(t, s"_${i + 1}")
-        }
-      def reductions(ast: Ast, path: List[Int] = List()): List[(Ident, Ast)] =
-        ast match {
-          case ident: Ident =>
-            List(ident -> property(path))
-          case Tuple(elems) =>
-            elems.zipWithIndex.flatMap {
-              case (elem, idx) => reductions(elem, path :+ idx)
-            }
-          case other =>
-            c.fail(s"Please report a bug. Expected tuple or ident, got '$other'")
-        }
-      BetaReduction(body, reductions(fields): _*)
+    case `patMatchParser`(value)            => value
+    case `blockParser`(block)               => block
+  }
+
+  val blockParser: Parser[Block] = Parser[Block] {
+    case q"{..$exprs}" if exprs.size > 0 => Block(exprs.map(astParser(_)))
+  }
+
+  val valParser: Parser[Val] = Parser[Val] {
+    case q"val $name: $typ = $body" => Val(ident(name), astParser(body))
+  }
+
+  val patMatchValParser: Parser[Val] = Parser[Val] {
+    case q"$mods val $name: $typ = ${ patMatchParser(value) }" =>
+      Val(ident(name), value)
+  }
+
+  val patMatchParser: Parser[Ast] = Parser[Ast] {
+    case q"$expr match { case ($fields) => $body }" =>
+      patMatchParser(expr, fields, body)
+  }
+
+  private def patMatchParser(tupleTree: Tree, fieldsTree: Tree, bodyTree: Tree) = {
```

Comment:
Cool, if patmatch is a common expression in eglish, I'm quite fine with
this. Thanks for the explanation.

Em ter, 15 de mar de 2016 19:49, godenji <notifications@github.com>
escreveu:

> In quill-core/src/main/scala/io/getquill/quotation/Parsing.scala
> <https://github.com/getquill/quill/pull/271#discussion_r56254723>:
>
> > +
> > +  val valParser: Parser[Val] = Parser[Val] {
> > +    case q"val $name: $typ = $body" => Val(ident(name), astParser(body))
> > +  }
> > +
> > +  val patMatchValParser: Parser[Val] = Parser[Val] {
> > +    case q"$mods val $name: $typ = ${ patMatchParser(value) }" =>
> > +      Val(ident(name), value)
> > +  }
> > +
> > +  val patMatchParser: Parser[Ast] = Parser[Ast] {
> > +    case q"$expr match { case ($fields) => $body }" =>
> > +      patMatchParser(expr, fields, body)
> > +  }
> > +
> > +  private def patMatchParser(tupleTree: Tree, fieldsTree: Tree, bodyTree: Tree) = {
>
> I abbreviated because in English you can say "patMatch" and programmers
> will understand "patternMatch". Plus case q"$expr match { case ($fields)
> => $body }" => pretty much says what it does ;-)
>
> Open to changing it though, let's see what @getquill/maintainers
> <https://github.com/orgs/getquill/teams/maintainers> think.
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/getquill/quill/pull/271/files/e9c942332120bd97748f207d6bc99c2ca2dc2ac0#r56254723>
>